From: SeongJae Park <sj@kernel.org>
Date: Tue, 7 Oct 2025 15:54:27 -0700
Subject: [PATCH] mm/damon/core: reduce rcu critical section for
 mem_cgroup_from_id()

The RCU critical section is necessary for only mem_cgroup_from_id().
And mem_cgroup_flush_stats() is sleepable, since css_rstat_flush(),
which is called by it, is sleepable.  Make it out of the RCU critical
section.

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index 65da79451bdb..3f50fdc74223 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -2140,13 +2140,13 @@ static unsigned long damos_get_node_memcg_used_bp(
 
 	rcu_read_lock();
 	memcg = mem_cgroup_from_id(goal->memcg_id);
+	rcu_read_unlock();
 	mem_cgroup_flush_stats(memcg);
 	lruvec = mem_cgroup_lruvec(memcg, NODE_DATA(goal->nid));
 	used_pages = lruvec_page_state(lruvec, NR_ACTIVE_ANON);
 	used_pages += lruvec_page_state(lruvec, NR_INACTIVE_ANON);
 	used_pages += lruvec_page_state(lruvec, NR_ACTIVE_FILE);
 	used_pages += lruvec_page_state(lruvec, NR_INACTIVE_FILE);
-	rcu_read_unlock();
 
 	si_meminfo_node(&i, goal->nid);
 	if (goal->metric == DAMOS_QUOTA_NODE_MEMCG_USED_BP)
-- 
2.39.5

